pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,189 @@
+type metadataPlugin struct {
+	host volume.VolumeHost
+}
+
+func (plugin *metadataPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *metadataPlugin) Name() string {
+	return metadataPluginName
+}
+
+func (plugin *metadataPlugin) CanSupport(spec *volume.Spec) bool {
+	return spec.VolumeSource.Metadata != nil
+}
+
+func (plugin *metadataPlugin) NewBuilder(spec *volume.Spec, podRef *api.ObjectReference, opts volume.VolumeOptions) (volume.Builder, error) {
+	return plugin.newBuilderInternal(spec, podRef, opts)
+}
+
+func (plugin *metadataPlugin) newBuilderInternal(spec *volume.Spec, podRef *api.ObjectReference, opts volume.VolumeOptions) (volume.Builder, error) {
+	return &metadataVolume{spec.Name,
+		spec.VolumeSource.Metadata.PodNameFileName,
+		spec.VolumeSource.Metadata.NamespaceFileName,
+		spec.VolumeSource.Metadata.LabelsFileName,
+		spec.VolumeSource.Metadata.AnnotationsFileName,
+		*podRef, plugin, &opts}, nil
+}
+
+func (plugin *metadataPlugin) NewCleaner(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID)
+}
+
+func (plugin *metadataPlugin) newCleanerInternal(volName string, podUID types.UID) (volume.Cleaner, error) {
+	return &metadataVolume{volName,
+		"",
+		"",
+		"",
+		"",
+		api.ObjectReference{UID: podUID}, plugin, nil}, nil
+}
+
+// metadataVolume handles retrieving metadata from the API server
+// and placing them into the volume on the host.
+type metadataVolume struct {
+	volName             string
+	podNameFileName     string
+	namespaceFileName   string
+	labelsFileName      string
+	annotationsFileName string
+	podRef              api.ObjectReference
+	plugin              *metadataPlugin
+	opts                *volume.VolumeOptions
+}
+
+func (m *metadataVolume) SetUp() error {
+	return m.SetUpAt(m.GetPath())
+}
+
+// formatMap formats labels from map[string]string in a string.
+func formatMap(labelMap map[string]string) string {
+	var l string
+	for key, value := range labelMap {
+		l += key + "=" + value + "\n"
+	}
+	return l
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageTypeMemory}},
+}
+
+func (m *metadataVolume) writeStringInFile(fileName, s string) error {
+	if s == "" || fileName == "" {
+		return nil
+	}
+	fullFilePath := path.Join(m.GetPath(), fileName)
+	fmt.Printf("---------- metatadaVolume.writeStringInFile '%s', '%s'\n", fullFilePath, s)
+	if err := ioutil.WriteFile(fullFilePath, []byte(s), 0644); err != nil {
+		return err
+	}
+	return nil
+
+}
+
+func (m *metadataVolume) SetUpAt(dir string) error {
+	glog.V(3).Infof("Setting up a metadata volume %v for pod %v at %v", m.volName, m.podRef.UID, dir)
+
+	// Wrap EmptyDir,let it do the setup.
+	wrapped, err := m.plugin.host.NewWrapperBuilder(wrappedVolumeSpec, &m.podRef, *m.opts)
```

Comment:
SetUp() needs to be idempotent, and I don't see that here.  In fact Secret seems to have the same problem.  @pmorie am I missing something?  It looks like every 10 seconds (or something like that) we will mount a new tmpfs over the old one(s).  Can you please verify, and if there is something I am missing that does make it idempotent, document it?

@vishh this is something to watch for when reviewing volume plugins.